### ISSUE TITLE
ML8 uses declarative rewriter..

### ISSUE THREAD

**Issue by grtjn** [ryan321/rest-cookies#1]
The approach to leverage rewriter.xqy is a little fragile, and could well break in latest MarkLogic as it has moved forward to using the declarative rewriter.xml.

Instead of having this rewriter in between, you could also let the middle-tier take care of that part. The middle-tier will have to respond to 401 anyhow, to confront the end-user with a login screen.

Or am I overlooking something?

**Comment by ryan321** [comment]
The rewriter.xqy is a modification of the original global rewriter for the REST API. The changes are to insert a hook to be able to perform authentication from a cookie since Basic authentication is not being used, and so no authentication has been done yet on the ML request. The middle-tier can't do this because the xdmp:login needs to happen in the context of the request on ML.

**Comment by grtjn** [comment]
I could be mistaken, but xdmp:login already creates a cookie, that is detected automatically by MarkLogic. Then again, you are storing sessions manually also (to not depend on sticky mode of load balancers), so maybe you do need to force a login before the actual REST api execution..

The issue is though that the REST api is no longer using the original global rewriter. It is using the declarative one instead since 8. Even relying on rest endpoints config directly (already no longer in use) is likely to be not future proof. I filed an RFE to be able to extend upon the declarative rewriter, but until then, this is a bit awkward. Roxy hybrid mode has a similar issue.. :(

**Comment by ryan321** [comment]
Yes the cookie from ML could be used, but it is tied to the enode so you'd have to use smart routing to always route the user's request to the same enode, and it is only in memory and not durable, meaning if the enode is restarting the session is lost. But it is still an option.

As for using rewriter.xml, I am for that. I just didn't try to use it.
